### ISSUE TITLE
proposal: `no_list_creation_for_typed_list`

### ISSUE THREAD

**Issue by dnfield** [dart-lang/linter#3106]
Context at https://github.com/flutter/engine/pull/30293

BAD:

```dart
class Foo {
  const Foo(this.a, this.b, this.c, this.d);

  final double a;
  final double b;
  final double c;
  final double d;

  // Intermediate allocation that will be discarded uses extra space and time.
  Float32List get typedList => Float32List.fromList(<int>[a, b, c, d]);
}
```

GOOD

```dart
class Foo {
  const Foo(this.a, this.b, this.c, this.d);

  final double a;
  final double b;
  final double c;
  final double d;

  Float32List get typedList {
    final Float32List typedList =  Float32List(4);
    typedList[0] = a;
    typedList[1] = b;
    typedList[2] = c;
    typedList[3] = d;
    return typedList;
  }
}
```

@rakudrama has a bug somewhere about making cheaper initializers for lists that could be elided for this kind of case, but I can't seem to find it. @mraleph might also find this interesting.

**Comment by dnfield** [comment]
Another thing that could help this case is if `Float32List` and `Float64List` had const constructors.

**Comment by dnfield** [comment]
It's used in `Rect` and `RRect` code, which is called frequently during frameworkload.

See https://github.com/flutter/engine/pull/30293

**Comment by mraleph** [comment]
`Float32List.fromList(<double>[...])` is especially bad because it boxes all the floating point values for no good reason. We should definitely consider special casing this. /cc @alexmarkov @mkustermann

**Comment by dnfield** [comment]
https://github.com/dart-lang/sdk/issues/44389 is the issue I couldn't find

**Comment by alexmarkov** [comment]
https://dart-review.googlesource.com/c/sdk/+/175821 added specialized `_GrowableList._literalN(a<0>, ..., a<N-1>)` constructors for list literals of N elements (where N=1..8). That saved a bit of space.

We can probably match pattern `Float32List|Float64List.from(<double>[a<0>, ..., a<N-1>])` and replace it with a specialized constructor for Float32List / Float64List in a similar way. But maybe if we add such public constructors for N elements in Float32List / Float64List, we could just write this code with new constructors directly.

**Comment by rakudrama** [comment]
If a special optimization is implemented for VM, something similar should happen in dart2js for Flutter-web.

The advantage of (public) constructors for up to N elements is that the lowering can happen in common code shared by DDC/dart2js/VM. The disadvantage is there are 9 typed arrays * N arguments is a lot of bloat for DDC, which can't tree-shake code due to separate compilation.

**Comment by rakudrama** [comment]
I'm not sure a similar optimization would be a benefit in all cases for dart2js.
Example: 
https://cs.opensource.google/flutter/engine/+/master:lib/web_ui/lib/src/engine/canvaskit/painting.dart;l=296?q=%22_invertColorMatrix%22&ss=flutter
Since this code is executed at most once, it would be best to try to compile it to
```
new Float32Array([-1,0,0,1,0,0,-1,0,1,0,1,1,1,1,0])
```
Anything else is likely to be larger.

**Comment by pq** [comment]
On the surface, optimizing in the VM sounds attractive but I don't fully understand the trade-offs.

To @rakudrama's point:

> 9 typed arrays * N arguments is a lot

but I don't have a sense really for what that bloat would cost us.

That said, a lint would be straightforward though obviously it'd need to be socialized and opted into.

**Comment by dnfield** [comment]
@rakudrama - in Flutter Web's implementation of Rect/RRect there's no issue because it doesn't use Float32/64Lists to move data to other code (https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/ui/geometry.dart#L178).

**Comment by srawlins** [comment]
It sounds like this is better solved at compile-time than with a lint rule?
